aten/src/ATen/native/Loss.cpp

```diff
@@ -390,22 +390,28 @@ Tensor& mse_loss_backward_out(Tensor& grad_input, const Tensor& grad_output,
 }
 
 Tensor l1_loss(const Tensor& input, const Tensor& target, int64_t reduction) {
-  auto loss = input.sub(target).abs_();
-  return apply_loss_reduction(loss, reduction);
+  if (input.is_complex()) {
+    const auto float_type = c10::toValueType(input.scalar_type());
+    Tensor result = at::empty({0}, input.options().dtype(float_type));
+    return at::l1_loss_out(result, input, target, reduction);
+  }
+  Tensor result = at::empty({0}, input.options());
```

Comment:
this should go in an else branch:
```
Tensor result;
if (input.is_complex()) {
...
} else {
    result = at::empty({0}, input.options());
}
```

Reply:
oh wait just saw you have a return statement in the if condition. I still think it might be cleaner to change it to an if else statement with a common return

Reply:
Hmm, then you'd have to declare `result` before the if else. Otherwise it would go out of scope by the time you try to return it.

Reply:
is that necessarily cleaner? :P I feel like it could be good either way.

Reply:
wait if you 'declare' it before, you are technically doing an extra default initialization then copy assigning instead of simply copy initializing

Reply:
A nice trick for this that @ezyang showed me is to use a lambda:
```cpp
const auto float_type = [&]() {
  if (input.is_complex()) {
    return c10::toValueType(input.scalar_type());
  } else {
    return input.scalar_type();
  }
}();
Tensor result = at::empty({0}, input.options().dtype(float_type));
return at::l1_loss_out(result, input, target, reduction);
```

But beyond that, what happens if you call `c10::toValueType` on a non complex dtype? Is it just returned as-is? If so, you don't need branching in this function at all!

Reply:
Ahh you're right `c10::toValueType` does handle the non-complex dtype by just returning as-is.